x/gamm/keeper/pool_service_test.go

```diff
@@ -430,6 +430,86 @@ func (suite *KeeperTestSuite) TestExitPool() {
 	}
 }
 
+// TestJoinPoolExitPool_InverseRelationship tests that joining pool and exiting pool
+// guarantees same amount in and out
+func (suite *KeeperTestSuite) TestJoinPoolExitPool_InverseRelationship() {
+	testCases := []struct {
+		name             string
+		pool             balancertypes.MsgCreateBalancerPool
+		joinPoolShareAmt sdk.Int
+	}{
+		{
+			name: "pool with same token ratio",
+			pool: balancer.NewMsgCreateBalancerPool(nil, balancer.PoolParams{
+				SwapFee: sdk.ZeroDec(),
+				ExitFee: sdk.ZeroDec(),
+			}, []balancertypes.PoolAsset{
+				{
+					Weight: sdk.NewInt(100),
+					Token:  sdk.NewCoin("foo", sdk.NewInt(10000)),
+				},
+				{
+					Weight: sdk.NewInt(100),
+					Token:  sdk.NewCoin("bar", sdk.NewInt(10000)),
+				},
+			}, defaultFutureGovernor),
+			joinPoolShareAmt: types.OneShare.MulRaw(50),
+		},
+		{
+			name: "pool with different token ratio",
+			pool: balancer.NewMsgCreateBalancerPool(nil, balancer.PoolParams{
+				SwapFee: sdk.ZeroDec(),
+				ExitFee: sdk.ZeroDec(),
+			}, []balancertypes.PoolAsset{
+				{
+					Weight: sdk.NewInt(100),
+					Token:  sdk.NewCoin("foo", sdk.NewInt(7000)),
+				},
+				{
+					Weight: sdk.NewInt(100),
+					Token:  sdk.NewCoin("bar", sdk.NewInt(10000)),
+				},
+			}, defaultFutureGovernor),
+			joinPoolShareAmt: types.OneShare.MulRaw(50),
+		},
+	}
+
+	for _, tc := range testCases {
+		suite.SetupTest()
+
+		for _, acc := range suite.TestAccs {
+			suite.FundAcc(acc, defaultAcctFunds)
+		}
+
+		createPoolAcc := suite.TestAccs[0]
+		joinPoolAcc := suite.TestAccs[1]
+
+		// test account is set on every test case iteration, we need to manually update address for pool creator
+		tc.pool.Sender = createPoolAcc.String()
+
+		poolId, err := suite.App.GAMMKeeper.CreatePool(suite.Ctx, tc.pool)
+		suite.Require().NoError(err)
+
+		balanceBeforeJoin := suite.App.BankKeeper.GetAllBalances(suite.Ctx, joinPoolAcc)
+		fmt.Println(balanceBeforeJoin.String())
+
+		err = suite.App.GAMMKeeper.JoinPoolNoSwap(suite.Ctx, joinPoolAcc, poolId, tc.joinPoolShareAmt, sdk.Coins{})
+		suite.Require().NoError(err)
+
+		_, err = suite.App.GAMMKeeper.ExitPool(suite.Ctx, joinPoolAcc, poolId, tc.joinPoolShareAmt, sdk.Coins{})
+
+		balanceAfterExit := suite.App.BankKeeper.GetAllBalances(suite.Ctx, joinPoolAcc)
+		deltaBalance, _ := balanceBeforeJoin.SafeSub(balanceAfterExit)
```

Comment:
Does depositing and withdrawing tokens preserve the order in which they were initially deposited? If not, does multi-asset deposit require a specific order for the input address? If any order suffices, this test could be potentially be circumvented by depositing multiple tokens in at different weights in a different order than what the pool outputs upon withdrawal.

In any case, we should add brief test case(s) that cover deposits/withdrawals with shuffled order of input assets on pools that aren't 50/50

Reply:
wdym by "ordering", and "depositing" and "withdrawing' tokens?

Reply:
Oh nvm I think JoinPoolNoSwap handles ordering etc.

Reply:
Oh yeah, this is directly using the method `JoinPool` no need to worry about stuff here